docs/replica-sets.md

```diff
@@ -0,0 +1,159 @@
+# VirtualMachineReplicaSet
+
+## Overview
+
+In order to allow scaling up or down similar VMs, add a way to specify
+`VirtualMachine` templates and the amount of replicas required, to let the
+runtime create these `VirtualMachine`s.
+
+## Use-cases
+
+### Ephemeral VMs
+
+Scaling ephemeral VMs which only have read-only mounts, or work with a backing
+store, to keep temporary data, which can be deleted after the VM gets
+undefined.
+
+## Implementation
+
+A new object `VirtualMachineReplicaSet`, backed with a controller will be
+implemented:
+
+```yaml
+apiVersion: kubevirt.io/v1alpha1
+kind: VirtualMachineReplicaSet
+metadata:
+  name: myreplicaset
+spec:
+  replicas: 3
+  selector:
+    matchLabels:
+      mylabel: mylabel
+  template:
+    metadata:
+      name: test
+      labels:
+        mylabel: mylabel
+    spec:
+      domain:
+        devices:
+      [...]
+```
+
+`spec.template` is equal to a `VirtualMachineSpec`. `spec.replicas` specifies
+how many instances should be created out of `spec.template`. `spec.selector`
+contains selectors, which need to match `spec.template.metadata.labels`.
+
+The status looks like this:
+
+```yaml
+status:
+  conditions: null
+  replicas: 3
+  readyReplicas : 2
+```
+In case of a scaling error, a `ReplicaFailure` condition is added to the
+`status.conditions`. Further it shows the number of `VirtualMachine`s which
+are in a non-final state and which match `spec.selector` in the
+`status.replicas` field.  `status.readyReplicas` indicates how many of these
+replicas meet the ready condition.
+
+*Note* that at the moment when writing this proposal, there exist no
+readiness checks for VirtualMachines in Kubevirt. Therefore a `VirtualMachine` is
+considered to be ready, when reported by virt-handler as running or migrating.
+
+In case of a delete failure:
+
+```yaml
+status:
+  conditions:
+  - type: "ReplicaFailure"
+    status: True
+    reason: "FailureDelete"
+    message: "no permission to delete VMs"
+    lastTransmissionTime: "..."
+  replicas: 4
+  readyReplicas: 3
+```
+
+In case of a create failure:
+
+```yaml
+status:
+  conditions:
+  - type: "ReplicaFailure"
+    status: True
+    reason: "FailureCreate"
+    message: "no permission to create VMs"
+    lastTransmissionTime: "..."
+  replicas: 2
+  readyReplicas: 3
+```
+
+### Guarantees
+
+The VirtualMachineReplicaSet  does **not** guarantee that there will never be
```

Comment:
@smarterclayton Isn't this now pretty close to the guarantees a ReplicaSet gives? Maybe I just expressed it the wrong way. I wanted to weaken the guarantee to that of a ReplicaSet.

Reply:
The VMRS implementation behaves mostly like a RS now, but I see the potential for that to change. It seems possible we'll encounter use cases that require us to strengthen these guarantees in the future.  I'm doubtful we'll have to account for the same kind of strict ordering/persistence that SS aim to provide, but it's possible we'll have to diverge from the loose guarantees the k8s RS provides.

That said, at what point would our "enhanced" VMRS feature set confuse people that are already familiar with the k8s RS?  For me, as long as we plan to keep the VMRS focused on ephemeral VMs, I don't have a problem with drawing the comparison between VMRS and the k8s RS. The feature set between the two may diverge some, but conceptually they're both working with replicating ephemeral instances of immutable objects.